compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -174,68 +173,120 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     }
 
     fn check_argument_compat(
-        rust_abi: bool,
-        caller: TyAndLayout<'tcx>,
-        callee: TyAndLayout<'tcx>,
+        caller_abi: &ArgAbi<'tcx, Ty<'tcx>>,
+        callee_abi: &ArgAbi<'tcx, Ty<'tcx>>,
     ) -> bool {
-        if caller.ty == callee.ty {
-            // No question
+        // Heuristic for type comparison.
+        let layout_compat = || {
+            if caller_abi.layout.ty == callee_abi.layout.ty {
+                // No question
+                return true;
+            }
+            // Compare layout
+            match (caller_abi.layout.abi, callee_abi.layout.abi) {
+                // Different valid ranges are okay (once we enforce validity,
+                // that will take care to make it UB to leave the range, just
+                // like for transmute).
+                (abi::Abi::Scalar(caller), abi::Abi::Scalar(callee)) => {
+                    caller.value == callee.value
+                }
+                (
+                    abi::Abi::ScalarPair(caller1, caller2),
+                    abi::Abi::ScalarPair(callee1, callee2),
+                ) => caller1.value == callee1.value && caller2.value == callee2.value,
+                // Be conservative
+                _ => false,
+            }
+        };
+        // Padding must be fully equal.
+        let pad_compat = || {
+            if caller_abi.pad != callee_abi.pad {
+                trace!(
+                    "check_argument_compat: incompatible pad: {:?} != {:?}",
+                    caller_abi.pad,
+                    callee_abi.pad
+                );
+                return false;
+            }
             return true;
-        }
-        if !rust_abi {
-            // Don't risk anything
-            return false;
-        }
-        // Compare layout
-        match (caller.abi, callee.abi) {
-            // Different valid ranges are okay (once we enforce validity,
-            // that will take care to make it UB to leave the range, just
-            // like for transmute).
-            (abi::Abi::Scalar(caller), abi::Abi::Scalar(callee)) => caller.value == callee.value,
-            (abi::Abi::ScalarPair(caller1, caller2), abi::Abi::ScalarPair(callee1, callee2)) => {
-                caller1.value == callee1.value && caller2.value == callee2.value
+        };
+        // For comparing the PassMode, we allow the attributes to differ
+        // (e.g., it is okay for NonNull to differ between caller and callee).
+        // FIXME: Are there attributes (`call::ArgAttributes`) that do need to be checked?
```

Comment:
Most attributes are probably relevant, as they change where arguments are to be passed ("byval" is LLVM's way of indicating a stack copy), or some transformation done to the value, like the zero/sign-extensions.

Off the top of my head, something like `fn(i8)` and `fn(u8)` may not be compatible on some architectures, if, say, the body immediately casts the argument to `i32`, since the widening may actually happen on the caller's side (one of the more annoying ABI details I've come across - on CPUs, at least, GPUs are all of this and worse).

Though, to be fair, the `layout_compat` check above should disallow signedness changes.

I suppose we should have a higher-level version of attributes before lowering to LLVM - e.g. the `NonNull` should be generated from the validity ranges in the `Layout` `abi` field in `rustc_codegen_llvm`, and the zero/sign-extension could be dictated by `Direct` and `Pair` actually holding register sizes.

Reply:
I originally started out checking the attributes for equality, but then very quickly got failures because the NonNull flag differs -- which I assume is a false positive. So it seems like `ArgAttributes` mixes crucial ABI information with hints that do not actually affect the call ABI. :/ (And yes it also looks awfully LLVM-specific.)

Reply:
> "byval" is LLVM's way of indicating a stack copy

I changed `"byval"` to the `on_stack` field of the `Indirect` pass mode.

> or some transformation done to the value, like the zero/sign-extensions.

Argument extension is specified by `attrs.arg_ext`. `attrs.pointee_align` should probably be checked by miri against the actual pointer value and `attrs.pointee_size` may be checked if it isn't already done by a stacked borrows protector. For `attrs.regular` I think only `InReg` changes the ABI. The rest introduces UB that miri could potentially check, but doesn't have to match between caller and callee.

Reply:
Miri does check that UB, but in a different way -- there are `Retag` MIR statements for this purpose.

Reply:
Okay, the latest version should now compare the relevant ArgAttributes.